pkg/tfbridge/tests/provider_test.go

```diff
@@ -129,6 +130,363 @@ func TestInputsConfigModeEqual(t *testing.T) {
 	}
 }
 
+// TestStateFunc ensures that resources with a StateFunc set on their schema are correctly
+// handled. This includes ensuring that the PlannedPrivate blob is passed from
+// PlanResourceChange to ApplyResourceChange. If this is passed correctly, the provider
+// will see the original value of the field, rather than the value that was produced by
+// the StateFunc.
+func TestStateFunc(t *testing.T) {
+	t.Parallel()
+	crosstests.Create(t,
+		map[string]*schema.Schema{
+			"test": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				StateFunc: func(v interface{}) string {
+					return v.(string) + " world"
+				},
+			},
+		},
+		cty.ObjectVal(map[string]cty.Value{
+			"test": cty.StringVal("hello"),
+		}),
+		crosstests.InferPulumiValue(),
+	)
+}
+
+// Regression test for [pulumi/pulumi-terraform-bridge#1767]
+func TestInputsUnspecifiedMaxItemsOne(t *testing.T) {
+	t.Parallel()
+	crosstests.Create(t,
+		map[string]*schema.Schema{
+			"f0": {
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"x": {Optional: true, Type: schema.TypeString},
+					},
+				},
+			},
+		},
+		cty.ObjectVal(map[string]cty.Value{}),
+		crosstests.InferPulumiValue(),
+	)
+}
+
+// Regression test for [pulumi/pulumi-terraform-bridge#1970] and [pulumi/pulumi-terraform-bridge#1964]
+func TestOptionalSetNotSpecified(t *testing.T) {
+	t.Parallel()
+	crosstests.Create(t,
+		map[string]*schema.Schema{
+			"f0": {
+				Optional: true,
+				Type:     schema.TypeSet,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"x": {Optional: true, Type: schema.TypeString},
+					},
+				},
+			},
+		},
+		cty.ObjectVal(map[string]cty.Value{}),
+		crosstests.InferPulumiValue(),
+	)
+}
+
+// Regression test for [pulumi/pulumi-terraform-bridge#1915]
+func TestInputsEqualEmptyList(t *testing.T) {
+	t.Parallel()
+	for _, maxItems := range []int{0, 1} {
+		t.Run(fmt.Sprintf("MaxItems: %v", maxItems), func(t *testing.T) {
+			crosstests.Create(t,
+				map[string]*schema.Schema{
+					"f0": {
+						Optional: true,
+						Type:     schema.TypeList,
+						MaxItems: maxItems,
+						Elem: &schema.Resource{
+							Schema: map[string]*schema.Schema{
+								"x": {Optional: true, Type: schema.TypeString},
+							},
+						},
+					},
+				},
+				cty.ObjectVal(map[string]cty.Value{
+					"f0": cty.ListValEmpty(cty.String),
+				}),
+				crosstests.InferPulumiValue(),
+			)
+		})
+	}
+}
+
+// TestAccCloudWatch failed with PlanResourceChange to do a simple Create preview because the state upgrade was
+// unexpectedly called with nil state. Emulate this here to test it does not fail.
+func TestCreateDoesNotPanicWithStateUpgraders(t *testing.T) {
```

Comment:
It is 100% surprising that state upgrade is at all called in Create. A comment should probably address this. Do we believe that TF calls state upgraders on Create an that is why Pulumi does it?

Reply:
New test looks like this:

```go
// TestCreateDoesNotInvokeStateUpgraders ensures that state upgrade machinery is not
// invoked during Create operations.
func TestCreateDoesNotInvokeStateUpgraders(t *testing.T) {
	t.Parallel()
	resource := func() *schema.Resource {
		return &schema.Resource{Schema: map[string]*schema.Schema{
			"f0": {
				Type:     schema.TypeString,
				Optional: true,
			},
		}}
	}

	upgradeFunc := func(ctx context.Context, rawState map[string]any, meta any) (map[string]any, error) {
		panic("State upgraders should not be called during create")
	}

	crosstests.Create(t,
		resource().Schema,
		cty.ObjectVal(map[string]cty.Value{
			"f0": cty.StringVal("default"),
		}),
		crosstests.InferPulumiValue(),
		crosstests.CreateStateUpgrader(1, []schema.StateUpgrader{
			{
				Type:    resource().CoreConfigSchema().ImpliedType(),
				Upgrade: upgradeFunc,
				Version: 0,
			},
		}),
	)
}
```